### ISSUE TITLE
unambiguous migration ordering during running migrations:down

### ISSUE THREAD

**Issue by peteraba** [opulencephp/Opulence#106]
### Issue description
The `executedmigrations` table will contain the datetime of the execution in the `daterun` column and this would be then used for finding the latest executed migration if you ran `./apex migrations:down`. The problem is, if you ran previous migrations at once before, this will not provide a sufficient way to select the latest migration run. Instead it will pick a random one executed in the exact second of the truly last executed one.

#### Code

```php
namespace AbterPhp\Users\Databases\Migrations;
// ...
class Init extends Migration
{
    // ...
    public static function getCreationDate(): DateTime
    {
        return DateTime::createFromFormat(DateTime::ATOM, '2019-02-28T21:00:00+00:00');
    }
    // ...
}
```
```php
namespace AbterPhp\Pages\Databases\Migrations;
// ...
class Init extends Migration
{
    // ...
    public static function getCreationDate(): DateTime
    {
        return DateTime::createFromFormat(DateTime::ATOM, '2019-02-28T21:01:00+00:00');
    }
    // ...
}
```
```php
namespace AbterPhp\Files\Databases\Migrations;
// ...
class Init extends Migration
{
    // ...
    public static function getCreationDate(): DateTime
    {
        return DateTime::createFromFormat(DateTime::ATOM, '2019-02-28T22:00:00+00:00');
    }
    // ...
}
```
```php
namespace AbterPhp\Contact\Databases\Migrations;
// ...
class Init extends Migration
{
    // ...
    public static function getCreationDate(): DateTime
    {
        return DateTime::createFromFormat(DateTime::ATOM, '2019-02-28T23:00:00+00:00');
    }
    // ...
}
```

#### Database

| migration | daterun |
--------------| -----------|
AbterPhp\Users\Databases\Migrations\Init | 2019-03-04 12:03:43 |
AbterPhp\Contact\Databases\Migrations\Init | 2019-03-04 12:03:44 |
AbterPhp\Files\Databases\Migrations\Init | 2019-03-04 12:03:44 |
AbterPhp\Pages\Databases\Migrations\Init | 2019-03-04 12:03:44 |

It may any of the last 3 for `migrate:down`.

### Incomplete list of potential fixes

1. Replace `daterun` column with a time format which will contain some high resolution time. (Backwards compatible, but requires a schema upgrade)
3. Store the `getCreationDate` in a column and use that as a second sorting parameter when selecting the "last run" migration.
2. Retrieve all migrations run at the same time as the "last" one and select the migration with the latest `getCreationDate` returned. (Backwards compatible and no schema upgrade necessary, but he query will be somewhat harder to understand, and could lead to problems if someone changes the `getCreatedAt` value of a migration.)
4. ~~Sleep 1 second before executing consecutive schema upgrades.~~ :smile: 

**Note:** Using microseconds as high resolution time is probably enough in most cases although it might just make it harder to reproduce the issue. Giving an option to use `hrtime` instead could have benefits, but only doable in PHP at the moment afaik. (https://pecl.php.net/package/hrtime, http://php.net/manual/en/function.hrtime.php) Without `hrtime` precision it might make sense to combine more solutions too.

**Comment by peteraba** [comment]
@davidbyoung if you could share what's on your mind about this, I could probably prepare a PR for you.

**Comment by Thijs-Riezebeek** [comment]
How about using an integer auto increment primary key for the migrations
table, that way order can be retrieved using the PK.

**Comment by davidbyoung** [comment]
I think the "right" way to do this would be using the approach @Thijs-Riezebeek suggested (auto-incrementing PK).  However, this would require updating schema, which isn't ideal.  We could at least update the [table creation query](https://github.com/opulencephp/Opulence/blob/1.1/src/Opulence/Framework/Databases/Migrations/SqlExecutedMigrationRepository.php#L128), but this would still require a manual change for anyone with existing migration tables.

**Comment by peteraba** [comment]
@davidbyoung I got to a point where I could start working on this, but I'd prefer to have my simpler PRs discussed and - ideally - merged first.

**Comment by peteraba** [comment]
@davidbyoung I'll pick this up now. Any preference for the name and type(s) of the primary key? (both MySQL and PostgreSQL)

**Comment by davidbyoung** [comment]
Probably "id".  It should be a sequential int.

**Comment by davidbyoung** [comment]
Fixed in v1.1.9.  Thanks for reporting and fixing the issue!
